src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -38,7 +37,7 @@ internal sealed class RSession : IRSession, IRCallbacks {
         public event EventHandler<EventArgs> Disconnected;
         public event EventHandler<EventArgs> Disposed;
         public event EventHandler<EventArgs> DirectoryChanged;
-        public event EventHandler<EventArgs> BeforePackagesInstalled;
+        public event AsyncEventHandler<EventArgs> BeforePackagesInstalledAsync;
```

Comment:
> OK. What's wrong with them?

- Async events don't support cancellation tokens. `CancellationToken` can be passed through as a property of `EventArgs`, but it isn't clear what to do if one of the handlers have thrown `OperationCanceledException` by itself.
- Sync event stops invocation when any of the handlers throw exception. Async handlers may or may not follow this rule. With more than one exception being produced during one invocation, we have to throw `AggregateException` which requires lots of code to handle. 
- `OperationCanceledException` vs regular exception. If event was canceled, should we still throw exception from one of the handlers or it should be swallowed?

And so on. Async event requires some good design before it can be widely used in application. Not to mention that people tend to forget to call `InvokeAsync` and use `Invoke` instead, and if there is only one handler, there will be no bug until the next handler is added.

> Do we want then to add a method to `IRSessionCallback/RSessionCallback` and call the index directly?

Yes, similar to plots.